ts/screens/authentication/IdpLoginScreen.tsx

```diff
@@ -111,127 +107,166 @@ const styles = StyleSheet.create({
-    this.props.dispatchLoginFailure(
-      new Error(`login failure with code ${errorCode || "n/a"}`),
-      this.idp
-    );
-    const logText = pipe(
-      errorCode,
-      O.fromNullable,
-      O.fold(
-        () => "login failed with no error code available",
-        ec =>
-          `login failed with code (${ec}) : ${getSpidErrorCodeDescription(ec)}`
-      )
-    );
-
-    handleSendAssistanceLog(this.choosenTool, logText);
-    this.setState({
-      requestState: pot.noneError(ErrorType.LOGIN_ERROR),
-      errorCode
-    });
+    setRequestState(pot.noneError(ErrorType.LOADING_ERROR));
   };
 
-  private handleLoginSuccess = (token: SessionToken) => {
-    handleSendAssistanceLog(this.choosenTool, `login success`);
-    this.props.dispatchLoginSuccess(token, this.idp);
+  const handleLoginFailure = useCallback(
+    (code?: string) => {
+      props.dispatchLoginFailure(
+        new Error(`login failure with code ${code || "n/a"}`),
+        idp
+      );
+      const logText = pipe(
+        code,
+        O.fromNullable,
+        O.fold(
+          () => "login failed with no error code available",
+          ec =>
+            `login failed with code (${ec}) : ${getSpidErrorCodeDescription(
+              ec
+            )}`
+        )
+      );
+
+      handleSendAssistanceLog(choosenTool, logText);
+      setRequestState(pot.noneError(ErrorType.LOGIN_ERROR));
+      setErrorCode(code);
+    },
+    [choosenTool, idp, props]
+  );
+
+  const handleLoginSuccess = (token: SessionToken) => {
+    handleSendAssistanceLog(choosenTool, `login success`);
+    props.dispatchLoginSuccess(token, idp);
   };
 
-  private setRequestStateToLoading = (): void =>
-    this.setState({ requestState: pot.noneLoading });
+  const setRequestStateToLoading = (): void => setRequestState(pot.noneLoading);
 
-  private handleNavigationStateChange = (event: WebViewNavigation): void => {
-    if (event.url) {
-      const urlChanged = getUrlBasepath(event.url);
-      if (urlChanged !== this.state.loginTrace) {
-        this.props.dispatchIdpLoginUrlChanged(urlChanged);
-        this.updateLoginTrace(urlChanged);
+  const handleNavigationStateChange = useCallback(
+    (event: WebViewNavigation) => {
+      if (event.url) {
+        const urlChanged = getUrlBasepath(event.url);
+        if (urlChanged !== loginTrace) {
+          props.dispatchIdpLoginUrlChanged(urlChanged);
+          updateLoginTrace(urlChanged);
+        }
       }
-    }
-    const isAssertion = pipe(
-      event.url,
-      O.fromNullable,
-      O.fold(
-        () => false,
-        s => s.indexOf("/assertionConsumerService") > -1
-      )
-    );
-    this.setState({
-      requestState:
+
+      if (publicKey && !lollipopCheck) {
+        const queryParam = new URLParse(event.url, true).query;
+        if (queryParam && queryParam.SAMLRequest) {
+          setWebviewSource(undefined);
+          const decodeSaml = decodeURIComponent(queryParam.SAMLRequest);
```

Comment:
Are we doing a double decoding on the `SAMLRequest`, here and inside `lollipopSamlVerify`?